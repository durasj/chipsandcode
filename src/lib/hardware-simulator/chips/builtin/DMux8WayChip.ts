import type { ChipPin } from '../Chip';
import type Chip from '../Chip';

/**
 * DMux8Way chip implemented natively
 */
class DMux8WayChip implements Chip {
  public readonly name = 'DMux8Way' as string;

  private in: boolean[];
  private sel: boolean[];
  private a: boolean[];
  private b: boolean[];
  private c: boolean[];
  private d: boolean[];
  private e: boolean[];
  private f: boolean[];
  private g: boolean[];
  private h: boolean[];

  constructor(private width = 1) {
    this.in = new Array(width).fill(false);
    this.sel = [false, false, false];
    this.a = new Array(width).fill(false);
    this.b = new Array(width).fill(false);
    this.c = new Array(width).fill(false);
    this.d = new Array(width).fill(false);
    this.e = new Array(width).fill(false);
    this.f = new Array(width).fill(false);
    this.g = new Array(width).fill(false);
    this.h = new Array(width).fill(false);
  }

  setInput(name: 'in' | 'sel', value: boolean[]) {
    this[name] = value;
  }

  getOutput(name: 'a' | 'b' | 'c' | 'd' | 'e' | 'f' | 'g' | 'h') {
    return this[name];
  }

  getPins() {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['in', { type: 'output', width: this.width, state: this.in, connections: [] }],
      ['sel', { type: 'input', width: 3, state: this.sel, connections: [] }],
      ['a', { type: 'input', width: this.width, state: this.a, connections: [] }],
      ['b', { type: 'input', width: this.width, state: this.b, connections: [] }],
      ['c', { type: 'input', width: this.width, state: this.c, connections: [] }],
      ['d', { type: 'input', width: this.width, state: this.d, connections: [] }],
      ['e', { type: 'input', width: this.width, state: this.e, connections: [] }],
      ['f', { type: 'input', width: this.width, state: this.f, connections: [] }],
      ['g', { type: 'input', width: this.width, state: this.g, connections: [] }],
      ['h', { type: 'input', width: this.width, state: this.h, connections: [] }],
    ]) satisfies Map<string, ChipPin>;
  }

  public run() {
    const selPin = this.selToPin();
    for (const pin of ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'] as const) {
      this[pin] = selPin === pin ? this.in : new Array(this.width).fill(false);
    }
  }

  private selToPin() {
    const sel = this.sel.map((v) => (v ? '1' : '0')).join('');

    switch (sel) {
      case '000':
        return 'a';
      case '001':
        return 'b';
      case '010':
        return 'c';
      case '011':
        return 'd';
      case '100':
        return 'e';
      case '101':
        return 'f';
      case '110':
        return 'g';
      case '111':
        return 'h';
      default:
        return 'a';
    }
  }
}

export default DMux8WayChip;
