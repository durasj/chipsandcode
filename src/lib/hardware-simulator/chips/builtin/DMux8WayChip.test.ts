import { describe, expect, it } from 'vitest';

import DMux8WayChip from './DMux8WayChip';

describe('DMux8Way Chip', () => {
  it('Provides interface info', async () => {
    const chip = new DMux8WayChip();

    expect(chip.name).toBe('DMux8Way');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new DMux8WayChip();

    const values = [
      [
        [false],
        [false, false, false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [false, false, false],
        [true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [false],
        [false, false, true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [false, false, true],
        [false],
        [true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [false],
        [false, true, false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [false, true, false],
        [false],
        [false],
        [true],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [false],
        [false, true, true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [false, true, true],
        [false],
        [false],
        [false],
        [true],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [false],
        [true, false, false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [true, false, false],
        [false],
        [false],
        [false],
        [false],
        [true],
        [false],
        [false],
        [false],
      ],
      [
        [false],
        [true, false, true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [true, false, true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [true],
        [false],
        [false],
      ],
      [
        [false],
        [true, true, false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [true, true, false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [true],
        [false],
      ],
      [
        [false],
        [true, true, true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
      ],
      [
        [true],
        [true, true, true],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [false],
        [true],
      ],
    ];

    values.forEach(([input, sel, a, b, c, d, e, f, g, h]) => {
      chip.setInput('in', input);
      chip.setInput('sel', sel);
      chip.run();

      expect(chip.getOutput('a')).toStrictEqual(a);
      expect(chip.getOutput('b')).toStrictEqual(b);
      expect(chip.getOutput('c')).toStrictEqual(c);
      expect(chip.getOutput('d')).toStrictEqual(d);
      expect(chip.getOutput('e')).toStrictEqual(e);
      expect(chip.getOutput('f')).toStrictEqual(f);
      expect(chip.getOutput('g')).toStrictEqual(g);
      expect(chip.getOutput('h')).toStrictEqual(h);
    });
  });
});
