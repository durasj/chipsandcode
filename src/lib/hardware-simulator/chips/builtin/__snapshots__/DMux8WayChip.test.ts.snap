// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DMux8Way Chip > Provides interface info 1`] = `
Map {
  "in" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
  "sel" => {
    "connections": [],
    "state": [
      false,
      false,
      false,
    ],
    "type": "input",
    "width": 3,
  },
  "a" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "b" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "c" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "d" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "e" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "f" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "g" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "h" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
}
`;
